File: internal/room/model/room.go
package model

import "time"

type Room struct {
	ID            uint        `gorm:"primaryKey" json:"id"`
	HotelID       uint        `gorm:"not null;index" json:"hotel_id"`
	RoomNumber    string      `gorm:"not null" json:"room_number"`
	Type          string      `gorm:"not null" json:"type"`
	Description   string      `json:"description"`
	PricePerNight float64     `gorm:"not null" json:"price_per_night"`
	Capacity      int         `gorm:"not null" json:"capacity"`
	Status        string      `gorm:"not null;default:'available'" json:"status"`
	Images        []RoomImage `gorm:"foreignKey:RoomID" json:"images"`
	CreatedAt     time.Time   `json:"created_at"`
	UpdatedAt     time.Time   `json:"updated_at"`

	// Not persisted - computed from Images after every fetch, mirroring
	// Hotel.ImageURL/ResolveImageURL exactly.
	ImageURL *string `gorm:"-" json:"image_url"`
}

func (r *Room) ResolveImageURL() {
	for _, img := range r.Images {
		if img.IsMain {
			url := img.URL
			r.ImageURL = &url
			return
		}
	}

	if len(r.Images) > 0 {
		url := r.Images[0].URL
		r.ImageURL = &url
	}
}

File: stayora-hotel-frontend/hotel-frontend/src/types/room.ts
// Mirrors the real GET /rooms/... response fields exactly (internal/room/model/room.go).

export type RoomStatus = "available" | "occupied" | "maintenance";

export interface RoomImage {
  id: number;
  room_id: number;
  url: string;
  is_main: boolean;
  created_at: string;
}

export interface Room {
  id: number;
  hotel_id: number;
  room_number: string;
  type: string;
  description: string;
  price_per_night: number;
  capacity: number;
  status: RoomStatus;
  images: RoomImage[] | null;
  image_url: string | null;
  created_at: string;
  updated_at: string;
}

// POST /hotels/:hotel_id/rooms and PUT /rooms/:id request body.
export interface RoomRequest {
  room_number: string;
  type: string;
  description?: string;
  price_per_night: number;
  capacity: number;
  status?: RoomStatus;
}
